### ISSUE TITLE
Working with PB files in C#

### ISSUE THREAD

**Issue by solarflarefx** [migueldeicaza/TensorFlowSharp#468]
Hello,

I am looking to train a model in TensorFlow with Python and then do inference in C#.  I used this MNist example as a guide: https://github.com/sdg002/MNISTpng

I am able to run all of the code in the example fine, but I do have some questions on the C# console application.  Within the "ConsoleAppTester" project, the code loads the pb model and then performed inference:
`graph.Import(buffer);
                using (var session = new TensorFlow.TFSession(graph))
                {
                    foreach (string file in filesTestingRandomized)
                    {
                        Stopwatch sw = new Stopwatch();
                        sw.Start();
                        var runner = session.GetRunner();
                        var tensor = Utils.ImageToTensorGrayScale(file);
                        runner.AddInput(graph["conv1_input"][0], tensor);
                        runner.Fetch(graph["activation_4/Softmax"][0]);

                        var output = runner.Run();
                        var vecResults = output[0].GetValue();
                        float[,] results = (float[,])vecResults;
                        sw.Stop();
                        ///
                        /// Evaluate the results
                        ///
                        int[] quantized = Utils.Quantized(results);
                        ///
                        /// Use the parent folder name to deterimine the expected digit
                        ///
                        string parentFolder = System.IO.Directory.GetParent(file).Name;
                        int iParentFolder = int.Parse(parentFolder);
                        int[] expected = Utils.GetQuantizedExpectedVector(iParentFolder);
                        bool success = quantized.SequenceEqual(expected);
                        if (!success) countOfFailedClassifications++;
                        string message = $"Directory={parentFolder}    File={System.IO.Path.GetFileName(file)}    Bit1={results[0, 0]} Bit2={results[0, 1]}   Elapsed={sw.ElapsedMilliseconds} ms, Success={success}";
                        Console.WriteLine(message);
                    }
                }
            }`

In particular, I have questions on the following line:
`runner.AddInput(graph["conv1_input"][0], tensor);
                        runner.Fetch(graph["activation_4/Softmax"][0]);`

Where are "conv1_input" and "activation_4/Softmax" defined in the pb file?  I believe these are the input and output layers of the network?  Where were these originally defined?  I am looking at "TrainedMnistFromFolder.py" and I see a "conv1" layer.  I also see activation layers defined, one of which is the "softmax" function.  But how do these become "conv1_input" and "activation_4/Softmax" when using inference in C#.

Is there a way to load a pb file and see these layers along with the names (or even get these by reading them in Python)?  Sorry I am relatively new working with pb files.  The key is understanding how to connect to the model at both input and output.

Thanks in advance.

**Comment by DrJasonhan** [comment]
You can use
` List<TFOperation> op_list = new List<TFOperation>(graph.GetEnumerator());`
to get all the layernames.
